hasktorch/src/Torch/Static/Native.hs

```diff
@@ -23,19 +23,1474 @@ import Data.Finite
+
+pow :: Scalar a => Tensor dtype shape -> a -> Tensor dtype shape
+pow t s = unsafePerformIO $ (cast2 ATen.pow_ts) t s
+
+relu :: Tensor dtype shape -> Tensor dtype shape
+relu t = unsafePerformIO $ (cast1 ATen.relu_t) t
+
+selu :: Tensor dtype shape -> Tensor dtype shape
+selu t = unsafePerformIO $ (cast1 ATen.selu_t) t
 
 sigmoid :: Tensor dtype shape -> Tensor dtype shape
-sigmoid t = UnsafeMkTensor $ D.sigmoid (toDynamic t)
+sigmoid t = unsafePerformIO $ (cast1 ATen.sigmoid_t) t
+
+sin :: Tensor dtype shape -> Tensor dtype shape
+sin t = unsafePerformIO $ (cast1 ATen.sin_t) t
+
+sinh :: Tensor dtype shape -> Tensor dtype shape
+sinh t = unsafePerformIO $ (cast1 ATen.sinh_t) t
+
+cos :: Tensor dtype shape -> Tensor dtype shape
+cos t = unsafePerformIO $ (cast1 ATen.cos_t) t
+
+sqrt :: Tensor dtype shape -> Tensor dtype shape
+sqrt t = unsafePerformIO $ (cast1 ATen.sqrt_t) t
+
+tanh :: Tensor dtype shape -> Tensor dtype shape
+tanh t = unsafePerformIO $ (cast1 ATen.tanh_t) t
+
+-- |
+-- >>> t = gt (ones :: Tensor Float '[2,2]) (zeros :: Tensor Float '[2,2])
+-- >>> (dtype t,shape t)
+-- (Bool,[2,2])
+gt :: Tensor dtype shape -> Tensor dtype shape -> Tensor Bool shape
+gt a b = unsafePerformIO $ (cast2 ATen.gt_tt) a b
+
+(>.) = gt
+
+lt :: Tensor dtype shape -> Tensor dtype shape -> Tensor Bool shape
+lt a b = unsafePerformIO $ (cast2 ATen.lt_tt) a b
+
+(<.) = lt
+
+ge :: Tensor dtype shape -> Tensor dtype shape -> Tensor Bool shape
+ge a b = unsafePerformIO $ (cast2 ATen.ge_tt) a b
+
+(>=.) = ge
+
+le :: Tensor dtype shape -> Tensor dtype shape -> Tensor Bool shape
+le a b = unsafePerformIO $ (cast2 ATen.le_tt) a b
+
+(<=.) = le
+
+-- |
+-- >>> t = eq (ones :: Tensor Float '[2,2]) (zeros :: Tensor Float '[2,2])
+-- >>> (dtype t,shape t)
+-- (Bool,[2,2])
+eq :: Tensor dtype shape -> Tensor dtype shape -> Tensor Bool shape
+eq a b = unsafePerformIO $ (cast2 ATen.eq_tt) a b
+
+(==.) = eq
+
+ne :: Tensor dtype shape -> Tensor dtype shape -> Tensor Bool shape
+ne a b = unsafePerformIO $ (cast2 ATen.ne_tt) a b
+
+(/=.) = ne
+
+-- |
+-- >>> t = (toDType (ones :: Tensor Float '[2,2]) :: Tensor Double '[2,2])
+-- >>> (dtype t,shape t)
+-- (Double,[2,2])
+toDType :: forall dtype dtype' shape. (Reifies dtype' DType) => Tensor dtype shape -> Tensor dtype' shape
+toDType t = unsafePerformIO $ (cast4 ATen.tensor_to_sbb) t (reflect (Proxy @dtype') :: DType) False False
+
+-- |
+-- >>> t = squeezeAll (ones :: Tensor Float '[2,1,2,1,2]) :: Tensor Float '[2,2,2]
+-- >>> (dtype t,shape t)
+-- (Float,[2,2,2])
+type family SqueezeAll (shape :: [Nat]) :: [Nat] where
+    SqueezeAll '[] = '[]
+    SqueezeAll (1: xs) = SqueezeAll xs
+    SqueezeAll (x: xs) = x ': SqueezeAll xs
+squeezeAll :: Tensor dtype shape -> Tensor dtype (SqueezeAll shape)
+squeezeAll t = unsafePerformIO $ (cast1 ATen.squeeze_t) t
+
+binary_cross_entropy_loss :: Tensor dtype shape -> Tensor dtype shape -> Tensor dtype shape -> Reduction-> Tensor dtype shape
```

Comment:
```suggestion
binary_cross_entropy :: Tensor dtype shape -> Tensor dtype shape -> Tensor dtype shape -> Reduction -> Tensor dtype '[]
```
see https://pytorch.org/docs/stable/nn.functional.html#binary-cross-entropy

Reply:
ah, it's more complicated than that. I just ran this, and this returns a `Tensor dtype shape` if the 4th argument is equal to `ReduceNone` and `Tensor dtype '[]` otherwise.

Reply:
I have implemented a shape chaging due to Reduction on type-level.
How about this one?
https://github.com/hasktorch/hasktorch/pull/163/commits/bc583bfc1c574e303b4fb8912716ed8f11a64429#diff-ed5b0b77220a9f439e61aa413e33bdfdR227-R266

Reply:
This is awesome!
it's a bit odd that you have to use a type application instead of a value for `reduction` now, but I don't see how to do it otherwise.